### PR TITLE
Fix issue 15309 again, without breaking the fix for regression 15550

### PR DESCRIPTION
**Pull request by 9rnsr**
[Issue 15309](https://issues.dlang.org/show_bug.cgi?id=15309) has fixed in #5263, but it had introduced a regression [issue 15550](https://issues.dlang.org/show_bug.cgi?id=15550).

By @MartinNowak, 15550 has been fixed in #5368, but it has also reopened 15309.

Fix 15309 again.

### CODE REVIEW

File: src/mtype.d [hunk 1]
```diff
@@ -7536,12 +7536,21 @@ public:
...
             if (sds.isPackage())
             {
                 error(loc, "%s has no type", exp.toChars());
                 goto Lerr;
             }
+            if (auto ti = sds.isTemplateInstance())
+            {
+                if (ti.needsTypeInference(sc))
+                {
+                    // Bugzilla 15550: ti is a partial instantiattion form fun!tiargs.
```
Comment (yebblies) [inline src/mtype.d]: Should be `instantiation`.

And would it be helpful to put some of this comment in the error message?  Maybe `expression (%s) is a partially instantiated template and has no type`?

Comment (9rnsr) [inline src/mtype.d]: Fixed typo,

The suggested message would be better, but diagnostic improvement is beyond the purpose of this PR.

I have a plan to move the "has no type" message into `ScopeExp.checkValue()`.

Comment (yebblies) [inline src/mtype.d]: Ok.  It would be nice to improve these errors as they can be unhelpful.

**Comment by yebblies** [comment]
Auto-merge toggled on

**Comment by 9rnsr** [comment]
Thanks!

File: src/expression.d [hunk 2]
```diff
@@ -5332,6 +5332,9 @@ public:
                         return e.semantic(sc);
                     }
                 }
+                // ti is an instance which requires IFTI.
+                sds = ti;
+                type = Type.tvoid;
```
Comment (MartinNowak) [inline src/expression.d]: This remains still problematic, b/c whenever we don't handle ScopeExp correctly, a wrong type (void) might slip through.
How about using `Type.terror` instead?

File: src/mtype.d [hunk 3]
```diff
@@ -7536,12 +7536,21 @@ public:
...
-            ScopeDsymbol sds = (cast(ScopeExp)exp).sds;
+            auto sds = (cast(ScopeExp)exp).sds;
             if (sds.isPackage())
             {
                 error(loc, "%s has no type", exp.toChars());
                 goto Lerr;
             }
+            if (auto ti = sds.isTemplateInstance())
+            {
+                if (ti.needsTypeInference(sc))
```
Comment (MartinNowak) [inline src/mtype.d]: Much better to handle this here, thx.

File: src/mtype.d (hunk 1 above)
Comment (9rnsr) [inline src/mtype.d]: By #5400, the error message for `typeof(func!tiargs)` will be improved.

**Pull request merged**